### DISCUSSION TITLE
Scan only included files and folders

### DISCUSSION THREAD

**Discussion by thumannw** [JetBrains/Qodana#154] [Q&A] [answered]
As far as I have understood from the documentation, you can only exclude paths from the scan. In our project, we only want to scan the java files of specific modules. Therefore, we have created a custom scope in IntelliJ, including only the Java folders of the respective modules. This works fine in IntelliJ. Is it possible somehow to reuse the scope file for the Qodana scan? If not, is therey any other way to achieve this with qodana.yaml? It is not feasible to exclude all the files except those Java files.

**Comment by brichbash** [comment, marked as answer]
Hello,

We're now working on documentation for the new feature - YAML profiles, which will allow you to use scopes similar to the ones IntelliJ has. It will become available within a month.
As for now, you can create your scope in IntelliJ, then [create an inspection profile](https://www.jetbrains.com/help/idea/customizing-profiles.html#create-inspection-profile) that uses the scope and after that use generated `.idea/inspectionProfiles/<your-profile-name>.xml` and `.idea/workspace.xml` from your project root.
For that to work, these files should be under your project root before Qodana analysis in your pipeline. Besides, you should [configure Qodana](https://www.jetbrains.com/help/qodana/qodana-yaml.html#Set+up+a+profile) to use the custom profile. 

Let me know if you have any questions.

**Reply by thumannw** [reply]
Thanks for the answer, the new feature sounds good. Let me just explain my motivation: We have been relying on IntelliJ's inspections and formatting for quite some while now and our rule was always 0 inspection results + everything formatted. Of course, without rigorous checking on CI, this was not always fulfilled. So our hope is that Qodana can fill this gap. However, because of the 0 inspection results rule, it is crucial that Qodana will reliably emit exactly the same inspection results as in the UI. I guess that the diverging scoping mechanisms is an obstacle for that.

Maybe the new feature will already be a solution for us, but anyway, I'd like to understand your proposed current solution a little better. I'm not quite sure how Qodana will pick up the scope defined in the UI. Is the information within the `workspace.xml` file? Since the latter also contains local information, I cannot check it into git. So which part of it contains the crucial information? I found the component "AnalysisUIOptions". Is that picked up by the Qodana run? Does that mean that if I run Qodana on a project for which I selected a given scope in the "Inspect Code..." dialog, the scan will always use that scope?

**Reply by brichbash** [reply]
Yes, the information about the scope is stored in `workspace.xml`. Here is an example of how the file can look like, it will be enough for Qodana to resolve the scope:
```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="NamedScopeManager">
    <scope name="Unnamed" pattern="file[project]:lib/utils/*||file[project]:lib/wrappers/*" />
  </component>
</project>
```
> Does that mean that if I run Qodana on a project for which I selected a given scope in the "Inspect Code..." dialog, the scan will always use that scope?

Not quite. The scope should be created in **Preferences | Appearance & Behaviour | Scopes**, and the scope should be set in your inspection profile. In **Preferences | Editor | Inspections** you can select a scope for each group of inspections. This information will be then stored in `.idea/inspectionProfiles/<your-profile-name>.xml`. After you set this up, Qodana will run selected inspections only in the specified scope.

**Reply by thumannw** [reply]
> Yes, the information about the scope is stored in `workspace.xml`. Here is an example of how the file can look like, it will be enough for Qodana to resolve the scope: ...

Ah now I understand. If you check "Share through VCS" in the window for defining the scope, it will be written to a separate file.

> Not quite. The scope should be created in **Preferences | Appearance & Behaviour | Scopes**, and the scope should be set in your inspection profile. In **Preferences | Editor | Inspections** you can select a scope for each group of inspections. This information will be then stored in `.idea/inspectionProfiles/<your-profile-name>.xml`. After you set this up, Qodana will run selected inspections only in the specified scope.

I see, that adds a lot of stuff to the profile xml. So maybe we wait for the new feature.

**Reply by thumannw** [reply]
Sorry @brichbash, I need more clarification for the new feature. Will it still be possible to reuse the profiles from the .idea folder? As pointed out above, the unique selling point of Qodana for us is that it brings the same results as IntelliJ via GUI. I guess the problem could already be solved by addding a new property to qodana.yaml where one can reference an existing scope, just as for the profile.

Another advantage of the selectable scope could be performance. Currently, it seems that Qodana scans each and every file it can in the project. This takes quite some time. If the scope is restricted with the new feature, will Qodana then only scan the included files?
